pkg/dyninst/ir/probe_definition.go

```diff
@@ -43,6 +50,32 @@ type Where interface {
 	Where() // marker method
 }
 
+// Segment is a segment of a probe template.
+type Segment interface {
+	Segment() // marker method
+}
```

Comment:
How about:

```
type TemplateSegment interface {
    AsStringSegment() (TemplateStringSegment, bool)
    AsExpressionSegment() (TemplateExpressionSegement, bool)
}

type TemplateStringSegment interface {
    TemplateSegment
    Value() string
}

type TemplateExpressionSegment interface {
    DSL() string
    AST() json.RawMessage
}
```

Also, I see no reason to include the index directly in the interface on probe definition. You will want it in whatever ends up in the ir.Program I suspect. 

This also lets us avoid duplicating interfaces in rcjson (which already imports ir)